src/main/java/rx/internal/operators/OperatorDoOnEmpty.java

```diff
@@ -0,0 +1,59 @@
+package rx.internal.operators;
+
+import rx.Observable;
+import rx.Subscriber;
+import rx.exceptions.Exceptions;
+import rx.functions.Action0;
+
+public final class OperatorDoOnEmpty<T> implements Observable.Operator<T, T> {
+
+    private final Action0 onEmpty;
+
+    public OperatorDoOnEmpty(Action0 onEmpty) {
+        this.onEmpty = onEmpty;
+    }
+
+    @Override
+    public Subscriber<? super T> call(final Subscriber<? super T> child) {
+
+        return new Subscriber<T>(child) {
+
+            private boolean isEmpty = true;
+            private boolean done = false;
+
+            @Override
+            public void onCompleted() {
+                if (done) {
+                    return;
+                }
+                if (isEmpty) {
+                    try {
+                        onEmpty.call();
+                    } catch (Exception e) {
+                        Exceptions.throwIfFatal(e);
```

Comment:
If a non-fatal error occurs in `onEmpty.call()` then at the moment it would complete normally. Moreover you should catch for `Throwable`s not just `Exception`s.

This should be 
```java 
catch (Throwable e) {
    Exceptions.throwOrReport(e, this);
    return;
}
```